packages/go_router/lib/src/match.dart

```diff
@@ -294,24 +667,110 @@ class RouteMatchList {
     );
     final Uri newUri =
         uri.replace(path: patternToPath(fullPath, newPathParameters));
-    return _copyWith(
+    return copyWith(
       matches: newMatches,
       uri: newUri,
       pathParameters: newPathParameters,
     );
   }
 
-  /// The last matching route.
-  RouteMatch get last => matches.last;
+  /// Returns a new List from the input matches with target removed.
+  ///
+  /// This method recursively looks into any ShellRouteMatch in matches and
+  /// removes target if it found a match in the match list nested in
+  /// ShellRouteMatch.
+  ///
+  /// This method returns a new list as long as the target is found in the
+  /// matches' subtree.
+  ///
+  /// If a target is found, the target and every node after the target in tree
+  /// order is removed.
+  static List<RouteMatchBase> _removeRouteMatchFromList(
+      List<RouteMatchBase> matches, RouteMatchBase target) {
+    // Remove is usually caused by pop; therefore, start searching from the end.
```

Comment:
>> Remove is usually caused by pop

by `usually` do you mean there are other cases?

this function looks like it assumes all removal is pop. it not only removes the target, but also remove all matches after the target.